shared/route-tree/render-route.js

```diff
@@ -0,0 +1,161 @@
+        routeLeafTags={leafTags || LeafTags()}
+        routeStack={stack || I.Stack()}
+        setRouteState={partialState => setRouteState(path, partialState)}
+      >{children}</RouteComponent>
+    )
+  }
+}
+
+type _RenderRouteProps = {
+  path: I.List<string>,
+} & RenderRouteProps
+
+type _RenderRouteResultParams = {
+  path: I.List<string>,
+  tags: LeafTags,
+  component: React$Element<any>,
+}
+
+const _RenderRouteResult: (spec?: _RenderRouteResultParams) => _RenderRouteResultParams & I.Record<_RenderRouteResultParams> = I.Record({
+  path: I.List(),
+  tags: LeafTags(),
+  component: null,
+})
+
+export type RouteRenderStack = I.Stack<_RenderRouteResult>
+
+// Render a route tree recursively. Returns a stack of rendered components from
+// the bottom (the currently visible view) up through each parent path.
+function _RenderRoute ({routeDef, routeState, setRouteState, path}: _RenderRouteProps): RouteRenderStack {
+  path = path || I.List()
+
+  if (!routeDef) {
+    throw new Error(`Undefined route: ${pathToString(path)}`)
+  } else if (!routeState) {
+    throw new Error(`Missing route state: ${pathToString(path)}`)
+  }
+
+  let stack
+  const selected = routeState.selected
+  if (selected === null) {
+    // If this is the current selected (bottom) view, initialize an empty
+    // stack. We'll add our view component to it as the first entry below.
+    if (!routeDef.component) {
+      throw new Error(`Attempt to render route without component: ${pathToString(path)}`)
+    }
+    stack = I.Stack()
+  } else {
+    // Otherwise, if this is a parent route, recurse to obtain the child stack.
+    let childDef = routeDef.getChild(selected)
+    const childState = routeState.children.get(selected)
+    const childPath = path.push(selected)
+    const childStack = _RenderRoute({routeDef: childDef, routeState: childState, path: childPath, setRouteState})
+
+    stack = childStack
+    if (routeDef.containerComponent) {
+      // If this route specifies a container component, compose it around every
+      // view in the stack.
+      stack = stack.map(r => (
+        r.update('component', child => (
+          <RenderRouteNode
+            isContainer={true}
+            routeDef={routeDef}
+            routeState={routeState}
+            path={path}
+            setRouteState={setRouteState}
+            leafTags={childStack.last().tags}
+            stack={childStack}
+          >
+            {child}
+          </RenderRouteNode>
+        ))
+      ))
+    }
+  }
+
+  if (routeDef.component) {
+    // If this path has a leaf component to render, add it to the stack.
+    const result = new _RenderRouteResult({
+      path,
+      component: (
+        <RenderRouteNode
+          isContainer={false}
+          routeDef={routeDef}
+          routeState={routeState}
+          path={path}
+          setRouteState={setRouteState}
+        />
+      ),
+      tags: routeDef.tags,
+    })
+    stack = stack.unshift(result)
```

Comment:
Hmm I'm not sure I agree! I'm thinking it's clearer to use `unshift`, which universally means "prepend", rather than `push`, which means "prepend because this is a Stack".